docs/11-MultiPageDashApp-LayoutFunctions.md

```diff
@@ -0,0 +1,134 @@
+
+```python
+import dash
+import dash_labs as dl
+import dash_bootstrap_components as dbc
+
+
+app = dash.Dash(
+    __name__,
+    plugins=[dl.plugins.pages],
+    external_stylesheets=[dbc.themes.BOOTSTRAP],
+)
+
+navbar = dbc.NavbarSimple(
+    dbc.Nav(
+        [
+            dbc.NavLink(page["name"], href=page["path"])
+            for page in dash.page_registry.values()
+            if page.get("top_nav")
+        ],
+    ),
+    brand="Multi Page App Demo",
+    color="primary",
+    dark=True,
+    className="mb-2",
+)
+
+app.layout = dbc.Container(
+    [navbar, dl.plugins.page_container],
+    fluid=True,
+)
+
+if __name__ == "__main__":
+    app.run_server(debug=True)
+
+```
+
+This is the `home.py` file (The `about.py` file is similar).  We are taking advantage of the ability to
+add arbitrary props to `dash.page_registry`.  We define `top_nav=True` for each page we want to include in the
+top nav.  See how this prop is used in `app.py` above when we create the top `navbar`.
+
+```python
+from dash import html
+import dash
+
+dash.register_page(__name__, top_nav=True)
+
+
+layout = html.Div("About page content")
+
+```
+
+
+We define the sidebar in `side_bar.py` in the `pages` folder.  Note that `sidebar` is a function (More on this later). It's
+defined here and imported into `topic_1.py, topic_2.py` and `topic_3.py`
+
+```python
+import dash
+from dash import html
+import dash_bootstrap_components as dbc
+
+
+def sidebar():
+    return html.Div(
+        dbc.Nav(
+            [
+                dbc.NavLink(
+                    [
+                        html.Div(page["name"], className="ms-2"),
+                    ],
+                    href=page["path"],
+                    active="exact",
+                )
+                for page in dash.page_registry.values()
+                if page["path"].startswith("/topic")
+            ],
+            vertical=True,
+            pills=True,
+            className="bg-light",
+        )
+    )
+
+```
+ Here is `topic_2.py`  (`topic_1.py` and `topic_3.py` are similar).  Note that the layout is also a function.
+
+```python
+from dash import html
+
+import dash
+import dash_bootstrap_components as dbc
+
+from .side_bar import sidebar
+
+dash.register_page(__name__)
+
+
+def layout():
+    return dbc.Row(
+        [dbc.Col(sidebar(), width=2), dbc.Col(html.Div("Topic 2 content"), width=10)]
+    )
+
+```
+
+The reason `sidebar` and the layouts for the three topic pages need to be functions is that data is added to
+`dash.register_page`as each module is imported from the `pages` folder and `dash.register_page` is called.
+If you don't use a function then all the pages may not yet be in `dash.page_registry` when the sidebar is created.
+When you use a function, the layout is created when it's used rather than when it's imported.
+
+Here is what the app looks like:
+
+![layout_functions](https://user-images.githubusercontent.com/72614349/147510416-3529dabd-6cf4-4e4f-b7a6-027267778b88.gif)
```

Comment:
Would it make sense to move this to the very top part of the page? I think the reader would appreciate being able to see what this chapter will teach them to create before they jump into all the reading.

Reply:
Oh, that would be a huge improvement --I'll make that change.  Thanks for your awesome feedback :-)